### DISCUSSION TITLE
OpenID Connect support

### DISCUSSION THREAD

**Discussion by nogweii** [healthchecks/healthchecks#752] [Ideas]
It would be nice if healthchecks supported OAuth/OpenID Connect or similar for on-premise deployments. That way a user could be redirected to a central IAM platform (AWS's, Active Directory's Federated Services, [Keycloak](https://www.keycloak.org/), there a number of similar providers and tools) that they likely are already logged in to.

**Comment by nogweii** [comment]
LDAP support (#198) is fairly similar to this request, though this is more about SSO -- redirecting the user to the authentication server rather than requiring them to type their username and password again.

**Comment by hongkongkiwi** [comment]
We would like to do Google Authentication, we already use this for all our other internal tools and it works really well taking the security burden away from us.

Can this be added? I think this would be invaluable for your amazing tool.

**Comment by jameskirsop** [comment]
Another way of approaching this would be to allow HTTP Headers to authenticate a user - in a similar way to Grafana or Zabbix (among many other tools) do.

Alternatively, I've got `django_auth_adfs` working quite well. I had to manually edit my superuser account to have my UPN as the username, but after that I'm able to sign in and get redirected to my projects with no issue. 

Adding ADFS support was pretty simple (in the same way that adding plain AD support seemed easy in #198). I just needed to make sure that `LOGIN_URL` was set so it redirected my login requests to ADFS instead of the standard `healthchecks.io` login page.

**Comment by nogweii** [comment]
@jameskirsop I would love to see any changes you've made to make ADFS work. Could you share those here, or perhaps as a gist?

**Comment by jameskirsop** [comment]
@evaryont, installed `django_auth_adfs` via pip and then used the following.

Added the appropriate config to `INSTALLED_APPS`, `MIDDLEWARE` and `AUTHENTICATION_BACKENDS`.

And used this:
```AUTH_ADFS = {
    "SERVER": "adfs.mycompany.com.au",
    "CLIENT_ID": "microsoft:identityserver:companyname-healthio",
    "RESOURCE": "companyname-healthio",
    "AUDIENCE": "microsoft:identityserver:companyname-healthio",
    "CLAIM_MAPPING": {"first_name": "given_name",
                      "last_name": "family_name",
                      "email": "email",
                      },
    "USERNAME_CLAIM": 'upn',
}

LOGIN_URL = "django_auth_adfs:login"
```

A couple of caveats I've found:
1. When inviting a user via email address the address is translated to lowercase. For users with capitals in their UPN, and then (in some cases) also in the email address handed back in claims, they're logged in with a new account instead because the two don't match. Eg. JBlogs@domain.com is stored as jblogs@domain.com and ADFS can't find a match, it seems.
 
2. If a new user logs in without having an account, the 'username' field is populated with their UPN. While this is expected behaviour with `django_auth_adfs`, it breaks functionality with Healthchecks - particularly when you try and invite an existing ADFS-created user to a project/team - and results in a 500 error - I'm expecting because of the first caveat:
```
django.urls.exceptions.NoReverseMatch: Reverse for 'hc-check-token' with arguments '('lowercaseaddress@lowercase.com', 'LLFZNDr4u9iGDFlOcb-TVHgIHj5GK2ZU')' not found. 1 pattern(s) tried: ['accounts\\\\/check_token\\\\/(?P<username>[-a-zA-Z0-9_]+)\\\\/(?P<token>[-a-zA-Z0-9_]+)\\\\/$']
```

The way I get around both of these, currently, is manually copying the token created by Healthcheck, deleting the 'invited' account created by Healthcheck (via `./manage.py shell`) and pasting the token into the username field in the database. It's a hacky way of doing it, but it seems to work.

Not using `LowercaseEmailField` for user accounts would be a step towards allowing better ADFS support in my opinion, but I do understand why it's a useful way of storing emails on public services. Alternatively, manually changing the way `django_auth_adfs` middleware handles this could be an appropriate option - to have it always translate email address claims to lowercase.

**Comment by jameskirsop** [comment]
A few hours later and I've made some significant progress with working around the above.

I've subclassed the `AdfsBaseBackend` and `AdfsAuthCodeBackend` classes from [django_auth_adfs](http://github.com/jobec/django-auth-adfs) and made a minor tweak so that it handles getting the appropriate claim back and then translates it to work correctly within the existing HealthChecks framework.

I'm going to do some more testing and then put together a PR for a secondary backend that can optionally be used to bridge the divide between HealthCheck and ADFS.

**Comment by lukasmrtvy** [comment]
References:
OIDC: https://django-oauth-toolkit.readthedocs.io/
LDAP: https://django-auth-ldap.readthedocs.io/en/latest/
ADFS: https://django-auth-adfs.readthedocs.io/en/latest/

**Comment by nogweii** [comment]
I have the ultimate hack, adding OpenID Connect support to Healthchecks without modifying any code:

```python
from .settings import INSTALLED_APPS as settings_ia
from .settings import AUTHENTICATION_BACKENDS as settings_ab
from .settings import ROOT_URLCONF as settings_ru

INSTALLED_APPS = list(settings_ia)
INSTALLED_APPS.insert(INSTALLED_APPS.index("django.contrib.auth")+1, "mozilla_django_oidc")
INSTALLED_APPS = tuple(INSTALLED_APPS)

AUTHENTICATION_BACKENDS = settings_ab + ("mozilla_django_oidc.auth.OIDCAuthenticationBackend",)

OIDC_RP_CLIENT_ID = "client-789"
OIDC_RP_CLIENT_SECRET = "secret-123"
OIDC_OP_AUTHORIZATION_ENDPOINT = "https://example.com/openid-connect/auth"
OIDC_OP_TOKEN_ENDPOINT = "https://example.com/openid-connect/token"
OIDC_OP_USER_ENDPOINT = "https://example.com/openid-connect/userinfo"
OIDC_OP_JWKS_ENDPOINT = "https://example.com/openid-connect/certs"
OIDC_RP_SIGN_ALGO = "RS256"

LOGIN_REDIRECT_URL = '/'

from django.urls import path, include
from django.views.generic.base import RedirectView
from types import ModuleType

# Create a dynamic module that injects some additional routes before HC's
sys.modules['sneaky_url'] = ModuleType('sneaky_url')
class _Sneaky(ModuleType):
  @property
  def urlpatterns(self):
    return [
      path('oidc/', include('mozilla_django_oidc.urls')),
      path('accounts/login/', RedirectView.as_view(url='/oidc/authenticate', permanent=True)),
      path('', include(settings_ru))
    ]
sys.modules['sneaky_url'].__class__ = _Sneaky
ROOT_URLCONF = 'sneaky_url'
```

Copy & paste this into `local_settings.py` and it should, roughly, Just Work. Oh, and make sure you run [`pip install mozilla-django-oidc`](https://github.com/mozilla/mozilla-django-oidc).

@cuu508 I'm sorry for this horrible python but it's pretty clever, eh?

**Reply by smkent** [reply]
@nogweii 2026 reporting in, this is still very useful! In particular the `ModuleType` dynamic module creation is clever! Always great to learn something.

I just set up Healthchecks for the first time, and I wanted to connect up to [Authelia](https://www.authelia.com/) for [OpenID Connect](https://www.authelia.com/configuration/identity-providers/openid-connect/provider/) (as I do most of my self hosted services). Since Healthchecks has its own authorization structure (e.g. the UI requires authentication while ping and API endpoints have their own authorization methods via keys and such), having OpenID Connect in Healthchecks directly avoids needing a bunch of `forward_auth` (or oauth2-proxy) URI exclusion patterns.

I modified your snippet with some new features:

* Automatic discovery of OIDC provider endpoint URLs from the provider URL's `/.well-known/openid-configuration` endpoint. For example, if your OIDC provider (e.g. Authelia) is at `https://login.example.com/`, you can just set the env var `OIDC_PROVIDER_URL`  to have the remaining endpoints discovered on Healthchecks startup. 
  * Some providers like Authentik and Keycloak may have per-client endpoints, such as `https://login.example.com/some/client/.well-known/openid-configuration`, in which case you'd set `OIDC_PROVIDER_URL` to `https://login.example.com/some/client`. I haven't used these, so YMMV.
* Accept some OIDC configuration via env variables (I'm running via Docker Compose so this is convenient)
* Add `mozilla_django_oidc.middleware.SessionRefresh` to `MIDDLEWARE` for [auth token refresh](https://mozilla-django-oidc.readthedocs.io/en/stable/installation.html#validate-id-tokens-by-renewing-them)

[Here's my updated `local_settings.py` snippet](https://github.com/smkent/homelab/commit/6bcf6c7e63a89bd579c061084863a06c9c3bd05a):

```python
# mypy: disable-error-code="import-not-found,import-untyped"
# ruff: noqa N811

import os
import sys
from collections.abc import Sequence
from types import ModuleType
from typing import Any

import requests
from django.urls import URLPattern, URLResolver, include, path
from django.views.generic.base import RedirectView

from .settings import AUTHENTICATION_BACKENDS as _auth_backends
from .settings import INSTALLED_APPS as _installed_apps
from .settings import MIDDLEWARE as _middleware
from .settings import ROOT_URLCONF as _root_urlconf
from .settings import envbool, envsecret

_oidc_provider_url = os.environ["OIDC_PROVIDER_URL"].removesuffix("/")
_oidc_conf = requests.get(
    f"{_oidc_provider_url}/.well-known/openid-configuration", timeout=5
).json()

OIDC_RP_CLIENT_ID = envsecret("OIDC_CLIENT_ID")
OIDC_RP_CLIENT_SECRET = envsecret("OIDC_CLIENT_SECRET")
OIDC_OP_AUTHORIZATION_ENDPOINT = _oidc_conf["authorization_endpoint"]
OIDC_OP_TOKEN_ENDPOINT = _oidc_conf["token_endpoint"]
OIDC_OP_USER_ENDPOINT = _oidc_conf["userinfo_endpoint"]
OIDC_OP_JWKS_ENDPOINT = _oidc_conf["jwks_uri"]
OIDC_RP_SIGN_ALGO = os.getenv("OIDC_RP_SIGN_ALGO", "RS256")
OIDC_RP_SCOPES = os.getenv("OIDC_RP_SCOPES", "openid email")
OIDC_TOKEN_USE_BASIC_AUTH = envbool("OIDC_TOKEN_USE_BASIC_AUTH", "False")
OIDC_CREATE_USER = envbool("OIDC_CREATE_USER", "True")

LOGIN_REDIRECT_URL = "/"


def _insert(src: Sequence[Any], val: Any, after: Any = None) -> tuple[Any]:
    src_list = list(src)
    where = (src_list.index(after) + 1) if after else len(src_list)
    src_list.insert(where, val)
    return tuple(src_list)


AUTHENTICATION_BACKENDS = _auth_backends + [
    "mozilla_django_oidc.auth.OIDCAuthenticationBackend"
]
INSTALLED_APPS = _insert(
    _installed_apps, "mozilla_django_oidc", "django.contrib.auth"
)
MIDDLEWARE = _insert(
    _middleware,
    "mozilla_django_oidc.middleware.SessionRefresh",
    "django.contrib.auth.middleware.AuthenticationMiddleware",
)


class _URLPatternsWithOIDC(ModuleType):
    @property
    def urlpatterns(self) -> list[URLResolver | URLPattern]:
        return [
            path("oidc/", include("mozilla_django_oidc.urls")),
            path(
                "accounts/login/",
                RedirectView.as_view(url="/oidc/authenticate", permanent=True),
            ),
            path("", include(_root_urlconf)),
        ]


ROOT_URLCONF = "hc.url_patterns_with_oidc"
sys.modules[ROOT_URLCONF] = _URLPatternsWithOIDC(ROOT_URLCONF)
```

In order to include [`mozilla-django-oidc`](https://github.com/mozilla/mozilla-django-oidc), [I build healthchecks with a small Dockerfile](https://github.com/smkent/homelab/blob/6bcf6c7e63a89bd579c061084863a06c9c3bd05a/compose/apps/healthchecks/build/Dockerfile) (pinned to Healthchecks `v4.0` instead of `latest`):

```docker
FROM healthchecks/healthchecks:v4.0
COPY *.py hc/
USER root
RUN pip install mozilla-django-oidc
USER hc
```

Drop these two files into a directory together with the appropriate environment configuration, and OpenID Connect should match users based on email address!

For bonus points, I wanted to have my Healthchecks instance match OpenID Connect users via their provider's static UUID instead of email address, so accounts will still be matched if a user changes their email address. To do so, subclassed `mozilla_django_oidc.auth.OIDCAuthenticationBackend`. First, I updated the class name in my above `local_settings.py`:

```diff
diff --git a/compose/apps/healthchecks/build/local_settings.py b/compose/apps/healthchecks/build/local_settings.py
index 6b23943..7ede6b8 100644
--- a/compose/apps/healthchecks/build/local_settings.py
+++ b/compose/apps/healthchecks/build/local_settings.py
@@ -43,9 +43,7 @@ def _insert(src: Sequence[Any], val: Any, after: Any = None) -> tuple[Any]:
     return tuple(src_list)
 
 
-AUTHENTICATION_BACKENDS = _auth_backends + [
-    "mozilla_django_oidc.auth.OIDCAuthenticationBackend"
-]
+AUTHENTICATION_BACKENDS = _auth_backends + ["hc.local.OIDCAuthSubClaimBackend"]
 INSTALLED_APPS = _insert(
     _installed_apps, "mozilla_django_oidc", "django.contrib.auth"
 )
```

and I then added my new `OIDCAuthSubClaimBackend` class in a new file `local.py` located in the same directory as `local_settings.py`:

```python
# mypy: disable-error-code="import-not-found,import-untyped"

from typing import Any

from hc.accounts.models import Profile
from hc.accounts.views import _make_user
from mozilla_django_oidc.auth import OIDCAuthenticationBackend


class OIDCAuthSubClaimBackend(OIDCAuthenticationBackend):
    def create_user(self, claims: dict[str, str]) -> Any:
        user = _make_user(claims["email"])
        profile = Profile.objects.for_user(user)
        profile.theme = "system"
        profile.save()
        return self.update_user(user, claims)

    def update_user(self, user: Any, claims: dict[str, str]) -> Any:
        if (sub := claims.get("sub")) and user.username != sub:
            user.username = sub
            user.save()
        is_admin = claims.get("healthchecks_role") == "admin"
        if is_admin != user.is_superuser or user.is_staff != user.is_superuser:
            user.is_staff = is_admin
            user.is_superuser = is_admin
            user.save()
        return user

    def filter_users_by_claims(self, claims: dict[str, str]) -> Any:
        if sub := claims.get("sub"):
            query = self.UserModel.objects.filter(username=sub)
            if query.count() > 0:
                return query
        return super().filter_users_by_claims(claims)
```

The core bit here is the `filter_users_by_claims` method; I believe this would still work without overriding `create_user` and `update_user`. These are further customizations for my setup that:

* Consume a custom `healthchecks` scope I configured in Authelia to indicate whether a user should have superuser privileges in Healthchecks, with permissions applied (or removed) on login.
* [Use Healthchecks' own `_make_user` helper for creating users](https://github.com/healthchecks/healthchecks/blob/v4.0/hc/accounts/views.py#L76-L107) instead of mozilla-django-oidc's user creation. This mimics the [user creation behavior Healthchecks uses when creating users via the custom header backend](https://github.com/healthchecks/healthchecks/blob/v4.0/hc/accounts/backends.py#L65-L92).
* Set the default UI theme for new users to `system` for automatic dark mode.

My full working setup is in my homelab repository, merged via https://github.com/smkent/homelab/pull/258, or [files browseable here in the merge commit](https://github.com/smkent/homelab/tree/118f7ecb481243f364c15a3693bc87972441287a/compose/apps/healthchecks). (`host.env` is just some custom glue to define `${DOMAIN}` and `${NODE_FQDN}` in the [container's environment](https://github.com/smkent/homelab/blob/118f7ecb481243f364c15a3693bc87972441287a/compose/apps/healthchecks/healthchecks.env) dynamically, not required for this solution).

**Comment by cuu508** [comment]
@evaryont that's quite a hack indeed! :+1:

I'm interested in adding "official" OpenID Connect support into the project and the hosted site as well. Picking the default identity providers will be tricky, but starting with Github and Google seems reasonable. 

Checking out mozilla-django-oidc library now – looks like it's designed to support a single, preconfigured identity provider, right?

**Comment by nogweii** [comment]
Indeed, looking around there isn't a way to configure it to support multiple providers simultaneously. You might be able to get it working by creating multiple applications, one for each provider, but boy that does sound hacky.

**Comment by decentral1se** [comment]
Is anyone working on this? I'd be most interested in Keycloak integration and I see that there is https://django-keycloak.readthedocs.io/en/latest/. Is anyone against me taking a stab at integrating this just for Keycloak?

**Comment by cuu508** [comment]
Hi @decentral1se – here's a quick summary of my plans and the current situation.

* I've looked at adding OIDC auth option. I haven't started any development on it. When looking at this, and making a survey of what other similar services support, it seemed Google and Github would be good candidates for initial identity providers to support. 
* I haven't looked at how OIDC auth would integrate the existing authentication – what new database fields, forms, email messages etc. would be needed. Some nontrivial work and design decisions here.
* I've also considered adding SAML support. No customer has explicitly asked for it, and there might be ways to support it indirectly via, say, Auth0. So I don't have near-term plans to work on it.

I'm not familiar with Keycloak (or, to be completely honest, with the other mentioned technologies), would Keycloak work as another OIDC identity provider? If that's the case, we should plan for supporting other OIDC providers in future too.

**Comment by lukasmrtvy** [comment]
Keycloak, https://github.com/dexidp/dex or generic oidc provider connector is always good start, because these will allow to You "federate" from other upstream oidc providers.

**Comment by lukasmrtvy** [comment]
@cuu508 is there any progress ?

**Comment by cuu508** [comment]
@lukasmrtvy from my side, unfortunately no.

**Comment by jameskirsop** [comment]
[Here's a blob of the subclasses](https://github.com/jameskirsop/healthchecks/blob/adfs_integration/hc/accounts/backends_adfs.py) I wrote for `django_auth_adfs` integration with Healthchecks. I've been using this in production for about 12 months now with no issue back against initially ADFS, and now AzureAD.

I don't know, @cuu508, if you'd like to have this in the project to allow for _some_ support of third party logins for those who are hosting themselves?? It might be a little niche, but it would solve the problem for a number of people who've posted on this issue.

I'm happy to create a PR for this after writing up some documentation, if you'd like to have it included.

**Comment by peppelinux** [comment]
I used djangosaml2 for doing SAML2 authentication, an example in my fork here, see saml2_sp and requirements:
https://github.com/UniversitaDellaCalabria/healthchecks

**Comment by Akruidenberg** [comment]
+1 for this feature!

**Comment by peppelinux** [comment]
I'd suggest also SATOSA as IAM Proxy

See
https://github.com/italia/Satosa-Saml2Spid

**Comment by nogweii** [comment]
With the HTTP header auth merged in, I think the only thing that remains would be an example configuration of something like the following stack:

* nginx/apache/traefik/caddy listening for incoming requests on 80 & 443
* [oauth2-proxy](https://oauth2-proxy.github.io/oauth2-proxy/) or [vouch-proxy](https://github.com/vouch/vouch-proxy) intercepting some portion of the requests (for example, we don't want to require an OIDC session to send a ping)
* Everything else being reverse proxied to django & Healthchecks itself.

Documenting that and referring people to it should hopefully provide the foundation for plugging in whatever authentication proxies people would like.

@cuu508 the only I'd like to ask of you is do you have a list of routes that should be excluded from from authentication? I could make some guesses, but it would be nice to be more confident about the choices.

**Comment by cuu508** [comment]
Hi @nogweii, here'sthe list of routes that I think would make sense to be excluded from authentication:

| urlpattern  | view  | view name |
| - | - | - |
| `/` | hc.front.views.index  | hc-index  |
| `/accounts/unsubscribe_reports/<str:signed_username>/`  | hc.accounts.views.unsubscribe_reports | hc-unsubscribe-reports  |
| `/api/v1/badges/` | hc.api.views.badges |   |
| `/api/v1/channels/` | hc.api.views.channels |   |
| `/api/v1/checks/` | hc.api.views.checks |   |
| `/api/v1/checks/<sha1:unique_key>`  | hc.api.views.get_check_by_unique_key  |   |
| `/api/v1/checks/<sha1:unique_key>/flips/` | hc.api.views.flips_by_unique_key  |   |
| `/api/v1/checks/<uuid:code>`  | hc.api.views.single | hc-api-single |
| `/api/v1/checks/<uuid:code>/flips/` | hc.api.views.flips_by_uuid  | hc-api-flips  |
| `/api/v1/checks/<uuid:code>/pause`  | hc.api.views.pause  | hc-api-pause  |
| `/api/v1/checks/<uuid:code>/pings/` | hc.api.views.pings  | hc-api-pings  |
| `/api/v1/metrics/`  | hc.api.views.metrics  |   |
| `/api/v1/notifications/<uuid:code>/status`  | hc.api.views.notification_status  | hc-api-notification-status  |
| `/api/v1/status/` | hc.api.views.status |   |
| `/badge/<slug:badge_key>/<slug:signature>.<slug:fmt>` | hc.api.views.badge  | hc-badge-all  |
| `/badge/<slug:badge_key>/<slug:signature>/<quoted:tag>.<slug:fmt>`  | hc.api.views.badge  | hc-badge  |
| `/checks/cron_preview/` | hc.front.views.cron_preview |   |
| `/docs/`  | hc.front.views.serve_doc  | hc-docs |
| `/docs/<slug:doc>/` | hc.front.views.serve_doc  | hc-serve-doc  |
| `/docs/cron/` | hc.front.views.docs_cron  | hc-docs-cron  |
| `/integrations/<uuid:code>/unsub/<str:signed_token>/` | hc.front.views.unsubscribe_email  | hc-unsubscribe-alerts |
| `/integrations/<uuid:code>/verify/<slug:token>/`  | hc.front.views.verify_email | hc-verify-email |
| `/integrations/add_pushover/` | hc.front.views.pushover_help  | hc-pushover-help  |
| `/integrations/add_slack/`  | hc.front.views.slack_help | hc-slack-help |
| `/integrations/pagerduty/`  | hc.front.views.pd_help  | hc-pagerduty-help |
| `/integrations/telegram/` | hc.front.views.telegram_help  | hc-telegram-help  |
| `/integrations/telegram/bot/` | hc.front.views.telegram_bot | hc-telegram-webhook |
| `/ping/<slug:ping_key>/<slug:slug>` | hc.api.views.ping_by_slug |   |
| `/ping/<slug:ping_key>/<slug:slug>/<int:exitstatus>`  | hc.api.views.ping_by_slug |   |
| `/ping/<slug:ping_key>/<slug:slug>/fail`  | hc.api.views.ping_by_slug |   |
| `/ping/<slug:ping_key>/<slug:slug>/start` | hc.api.views.ping_by_slug |   |
| `/ping/<uuid:code>` | hc.api.views.ping |   |
| `/ping/<uuid:code>/`  | hc.api.views.ping |   |
| `/ping/<uuid:code>/<int:exitstatus>`  | hc.api.views.ping |   |
| `/ping/<uuid:code>/fail`  | hc.api.views.ping |   |
| `/ping/<uuid:code>/start` | hc.api.views.ping |   |
| `/projects/<uuid:code>/checks/metrics/<slug:key>` | hc.front.views.metrics  |   |
| `/projects/<uuid:code>/metrics/<slug:key>`  | hc.front.views.metrics  | hc-metrics  |
| `/tv/`  | hc.front.views.dashboard  | hc-dashboard  |

**Comment by axelcypher** [comment]
> With the HTTP header auth merged in, I think the only thing that remains would be an example configuration of something like the following stack:
> 
> * nginx/apache/traefik/caddy listening for incoming requests on 80 & 443
> * [oauth2-proxy](https://oauth2-proxy.github.io/oauth2-proxy/) or [vouch-proxy](https://github.com/vouch/vouch-proxy) intercepting some portion of the requests (for example, we don't want to require an OIDC session to send a ping)
> * Everything else being reverse proxied to django & Healthchecks itself.
> 
> Documenting that and referring people to it should hopefully provide the foundation for plugging in whatever authentication proxies people would like.
> 
> @cuu508 the only I'd like to ask of you is do you have a list of routes that should be excluded from from authentication? I could make some guesses, but it would be nice to be more confident about the choices.

Soooo... did anyone managed to get it to work? Right now i'm trying to authenticate via Authentik with Traefik, but i can't get it to work...

**Reply by scheibling** [reply]
Recently set this up in Kubernetes, but should work on Docker as well (with the addition of a reverse proxy able to route things the way described in the ingress configuration, everything to the proxy except the /ping/ prefix) https://github.com/CloudyneS/healthchecks-k8s-oidc

**Comment by cuu508** [comment]
I currently have no plans to add OIDC authentication options. I want to preserve the discussion in the comments, so I'm converting this issue to a discussion rather than closing it.

**Comment by scheibling** [comment]
A very simple version of the oauth2-proxy/healthchecks setup (on k8s) is described here: https://github.com/CloudyneS/healthchecks-k8s-oidc. We recently set it up and it seems to be working well overall.

I have not specified the routes as explicitly as @cuu508 described, it's more of a working POC setup for adding authentication via OIDC that can be further customized. Will probably add those options at some point and convert it to a chart/something more complete and add a POC for docker as well.

**Comment by Coo-ops** [comment]
Need it, everything should have it. It's the future.

**Comment by Suertzz** [comment]
Hello guys,

Do you plan to add it near 2024 ? We need this feature 😇

**Reply by cuu508** [reply]
I currently have no plans to add OIDC authentication options.

If you are using a self-hosted instance, you can put an authentication proxy in front of Healthchecks. One example using oauth2-proxy was posted earlier in the comments – https://github.com/CloudyneS/healthchecks-k8s-oidc

IIUC some people are using Authelia with Healthchecks as well. I have no personal experience with either, so cannot give specific recommendations.

**Reply by TheCataliasTNT2k** [reply]
Would be great to have it...

**Comment by beardedsamwise** [comment]
I have SSO working with Okta using the OAuth2 Proxy (on AWS using ECS) without too much difficultly. I didn't need to modify the [uwsgi.ini](https://github.com/CloudyneS/healthchecks-k8s-oidc/blob/main/docker-image/uwsgi.ini) file as described in the example above. 

This [documentation](https://oauth2-proxy.github.io/oauth2-proxy/configuration/providers/openid_connect/) describes the config for the OAuth2 Proxy quite well. You simply need to configure the `REMOTE_USER_HEADER` as `HTTP_X_FORWARDED_EMAIL` on the Healthchecks container along with the all the OAuth2 Proxy configs and it works! 

If you have any questions regarding my implementation please reach out 😄

**Reply by ovizii** [reply]
> You simply need to configure the `REMOTE_USER_HEADER` as `HTTP_X_FORWARDED_EMAIL` on the Healthchecks container along with the all the OAuth2 Proxy configs and it works!

Would you mind explaining that please?

**Reply by beardedsamwise** [reply]
We set the `REMOTE_USER_HEADER` env var as `"HTTP_X_FORWARDED_EMAIL"` for the Healthchecks container, the rest of the SSO configs are handled by the OAuth 2 Proxy.

**Reply by ovizii** [reply]
Thanks, I got it working and then realized I need to create a second router with different rules, so I can let the pings go through without authentication.

**Reply by ovizii** [reply]
Btw. I gave up on this idea seeing this huge list of urlpattern I'd need to route past the authentication. Seemed way too much work :-)
https://github.com/healthchecks/healthchecks/discussions/752#discussioncomment-4420140

**Reply by beardedsamwise** [reply]
A lot of those can be wildcarded, the list we use isn't very long. Once we set up we've never had to touch it.

**Reply by nogweii** [reply]
@beardedsamwise could you share the oauth2-proxy config you're using?

**Reply by beardedsamwise** [reply]
We host Healthchecks on AWS ECS and configure the OAuth2 proxy using environment variables. Below is a snippet of our AWS CDK code:

Site root is the full root URL of your Healthchecks instance (ie. https://healthchecks.awesome.com)

The OIDC Issuer URL is whatever your Okta OAuth URL is (we have a dev and prod Okta environment). 

```
      environment: {
        OAUTH2_PROXY_COOKIE_REFRESH: "8h",
        OAUTH2_PROXY_PROVIDER: "oidc",
        OAUTH2_PROXY_COOKIE_SECURE: "true",
        OAUTH2_PROXY_EMAIL_DOMAINS: "*",
        OAUTH2_PROXY_REDIRECT_URL: `${siteRoot}/oauth2/callback`,
        // Generate a random cookie secret for production
        // In dev we use a fixed value so that we don't release a new ECS task definition every time we push to a branch
        OAUTH2_PROXY_COOKIE_SECRET: isProduction
          ? crypto.randomBytes(32 / 2).toString("hex")
          : "your_string_goes_here",
        OAUTH2_PROXY_PASS_ACCESS_TOKEN: "true",
        OAUTH2_PROXY_SKIP_PROVIDER_BUTTON: "true",
        OAUTH2_PROXY_OIDC_ISSUER_URL: props.sso.oktaIssuerUrl,
      },
 ```
 
 We also pass in the `OAUTH2_PROXY_CLIENT_ID` and `OAUTH2_PROXY_CLIENT_SECRET` as a secret from Secret Manager. 
 
 The container is started with:
 
 ```
       command: [
        "oauth2-proxy",
        "--http-address=0.0.0.0:4180",
        "--upstream=http://localhost:8000",
      ],
 ```
 
 The upstream is the actual Healthchecks container.

**Reply by nogweii** [reply]
So you don't skip authentication on any routes in oauth2-proxy? Are the list of routes configured to reach healthchecks directly via your load balancer?

**Reply by beardedsamwise** [reply]
I think these are the main ones we skip OAuth for:

`["/ping/*", "/api/*"]` and `["/badge/*", "/static/*"]`.

**Comment by Dr-Shadow** [comment]
I don't have any success using Zitadel with oauth2-proxy especially without a proper configuration for skip_auth_routes with the allowed methods as well.

OIDC support directly inside healthchecks would be a better bet I believe.

**Reply by ilyakonovalenko** [reply]
> OIDC support directly inside healthchecks would be a better bet I believe.

Agree. OIDC support is what I miss in HealthChecks.

**Comment by koehn** [comment]
FWIW if you're integrating behind an Authentik proxy you want to set the `REMOTE_USER_HEADER` environment variable to `HTTP_X_AUTHENTIK_EMAIL`. Works great!
